In [1]:
from datascience import *
import numpy as np
from math import *
from scipy import stats
import matplotlib.pyplot as plt
%matplotlib inline

## Lesson 32: Likelihood Ratio Tests

Last time, we introduced Likelihood Ratio tests. Recall that the point of a likelihood ratio test is to compare the likelihood function under a hypothesized value of the parameter with the liklihood function at its maximum. Instead of looking at the ratio $\Lambda$ itself, we often consider $-2\log \Lambda$ instead, since it has a handy distribution. 

### Example 1: Exponential Distribution

Suppose $X_1,X_2,...,X_n$ is an iid sequence of random variables from the exponential distribution with unknown parameter $\lambda$. Recall that the maximum likelihood estimate of $\lambda$ is $1\over\bar{X}$. We collect a random sample of size 20 and want to test the hypothesis $H_0: \lambda = 3$ vs $H_1: \lambda \neq 3$. Using the data in the python box below, conduct a likelihood ratio test on this hypothesis.  

In [2]:
my_data=np.array([0.18,0.277,0.105,0.126,0.225,0.026,0.123,0.423,0.006,0.281,0.050,0.692,0.105,0.275,0.346,0.079,0.045,0.222,0.063,0.281])

In [3]:
# Sole first for what Capital Lambda is by 
Xbar=np.mean(my_data)
lambdaML=1/Xbar
Caplambda=(3**20*exp(-20*3*Xbar))/(lambdaML**20*exp(-20*lambdaML*Xbar))
Estimator=-2*log(Caplambda)
1-stats.chi2.cdf(Estimator,1)

0.029827229194775096

Therefore, my p value is about .03 and since this is less than .05 we reject the null.

#### Power

Suppose that the true value of $\lambda$ is 5. Let's determine the power of this test. Let $n=20$. Then determine the power if $n=50$. Remember, power is the probability of correctly rejecting the null hypothesis. 

First, find what value of $-2 \log \Lambda$ would lead you to reject $H_0$. This is sometimes called the critical value. 

In [5]:
Crit_val=stats.chi2.ppf(.95,1)
Crit_val

3.841458820694124

Next, obtain the power. Obtain a sample of size 20 from the true population and obtain the value of $-2\log \Lambda$ for this sample. Repeat many times and determine how often you reject the null hypothesis. 

In [15]:
Array = []
exponent = stats.expon(scale = 1/5)
for i in np.arange(10000):
    A = exponent.rvs(size = 20)
    Xbar = np.mean(A)
    xsum = sum(A)
    Array = np.append(Array,2*(20*log(1/Xbar) - 20 - 20*log(3)+3*xsum))
    
power = sum(Array>Crit_val)/10000
power

0.5951

Repeated 10 times and always rejected the null hypothesis.

Repeat for a sample size of 50. What do you expect to happen to power? 

In [25]:
Array = []
exponent = stats.expon(scale = 1/5)
for i in np.arange(10000):
    A = exponent.rvs(size = 50)
    Xbar = np.mean(A)
    xsum = sum(A)
    Array = np.append(Array,2*(50*log(1/Xbar) - 50 - 50*log(3)+3*xsum))
    
power = sum(Array>Crit_val)/10000
power

0.9517

Repeated 10 times and rejected the null hypothesis once.

### A Different Test

We've explored hypothesis tests in this class before. Taking advantage of our computing power, we don't have to rely on test statistics with asymptotic distributions. Let's conduct a more direct hypothesis test using simulation. Recall:

$$
H_0: \lambda = 3
$$

$$
H_1: \lambda \neq 3
$$

Pick a different test statistic. Obtain an empirical distribution of that test statistic under $H_0$. Next, find the $p$-value by determining how often this test statistic is at or further away from the test statistic derived from the sample. Remember that this is a two-sided test. 

In [27]:
Array = []
exponent = stats.expon(scale = 1/3)
for n in np.arange(10000):
    X = exponent.rvs(size= 20)
    Array = np.append(Array,np.mean(X))
    
p = sum(Array>np.mean(my_data))/10000
1-p


0.01880000000000004

How did the $p$-value compare to the LRT $p$-value? I wonder how the power of this test compares to our LRT. 

Since 1.88% of the time will lambda=3 give a mean for a sample of size 20 that matches our data and 1.88<2.5 we reject the null hypothesis.

#### Power

Let's figure out the power of this test. First, determine for what values of the test statistic would lead us to reject $H_0$. These values can be referred to as your rejection region. 

In [28]:
Low,High=percentile([2.5,97.5],Array)
print(Low)
print(High)

0.20308841217019746
0.49339828956889675


Now, determine the power of this test. Like in the LRT case, obtain a sample of size 20 and obtain the test statistic. Repeat many times and see how often your test statistic is in your rejection region. 

In [37]:
Array = []
exponent = stats.expon(scale = 1/5)
for i in np.arange(10000):
    A = exponent.rvs(size = 20)
    Xbar = np.mean(A)
    Array = np.append(Array,Xbar)
    
power = sum(Array<Low)/10000 + sum(Array>High)/10000
power

0.5602

After 10 times, test statistic is always in the rejection region.

Repeat for a sample size of 50. Note that you will have to obtain new critical values in order to do this.  

In [42]:
Array = []
exponent = stats.expon(scale = 1/5)
for i in np.arange(10000):
    A = exponent.rvs(size = 50)
    Xbar = np.mean(A)
    Array = np.append(Array,Xbar)
    
power = sum(Array<Low)/10000 + sum(Array>High)/10000
power

0.5631

After 10 times, test statistic is always in the rejection region.